libkbfs/file_data.go

```diff
@@ -768,44 +1067,124 @@ func (fd *fileData) truncateExtend(ctx context.Context, size uint64,
 	}
 
+	dirtyMap := make(map[BlockPointer]bool)
 	if nextBlockOff > 0 {
-		// TODO: if indexInParent == 0, we can remove the level of indirection.
-		// TODO: support multiple levels of indirection.
-		parentBlock := parentBlocks[0].pblock
-		indexInParent := parentBlocks[0].childIndex
-		for _, ptr := range parentBlock.IPtrs[indexInParent+1:] {
-			unrefs = append(unrefs, ptr.BlockInfo)
+		// TODO: remove any unnecessary levels of indirection if the
+		// number of leaf nodes shrinks significantly (KBFS-1824).
+
+		// Get all paths to any leaf nodes following the new
+		// right-most block, since those blocks need to be
+		// unreferenced, and their parents need to be modified or
+		// unreferenced.
+		pfr, err := fd.getIndirectBlocksForOffsetRange(
+			ctx, topBlock, nextBlockOff, -1)
+		if err != nil {
+			return DirEntry{}, nil, nil, 0, err
 		}
-		parentBlock.IPtrs = parentBlock.IPtrs[:indexInParent+1]
-		// always make the parent block dirty, so we will sync it
-		if err = fd.cacher(fd.rootBlockPointer(), parentBlock); err != nil {
-			return DirEntry{}, nil, newlyDirtiedChildBytes, err
+
+		// A map from a pointer of an indirect block to that block's
+		// original set of block pointers, before they are truncated
+		// in the loop below.  It also tracks which pointed-to blocks
+		// have already been processed.
+		savedChildPtrs := make(map[BlockPointer][]IndirectFilePtr)
+		for _, path := range pfr {
+			// parentInfo points to pb.pblock in the loop below.  The
+			// initial case is the top block, for which we need to
+			// fake a block info using just the root block pointer.
+			parentInfo := BlockInfo{BlockPointer: fd.rootBlockPointer()}
+			leftMost := true
+			for i, pb := range path {
+				ptrs := savedChildPtrs[parentInfo.BlockPointer]
+				if ptrs == nil {
+					// Process each block exactly once, removing all
+					// now-unnecessary indirect pointers (but caching
+					// that list so we can still walk the tree on the
+					// next iterations).
+					pblock := pb.pblock
+					ptrs = pblock.IPtrs
+					savedChildPtrs[parentInfo.BlockPointer] = ptrs
+
+					// Remove the first child iptr and everything
+					// following it if all the child indices below
+					// this level are 0.
+					removeStartingFromIndex := pb.childIndex
+					for j := i + 1; j < len(path); j++ {
+						if path[j].childIndex > 0 {
+							removeStartingFromIndex++
+							break
+						}
+					}
+
+					// If we remove iptr 0, this block can be
+					// unreferenced (unless it's on the left-most edge
+					// of the tree, in which case we keep it around
+					// for now -- see above TODO).
+					if pb.childIndex == 0 && !leftMost {
+						if parentInfo.EncodedSize != 0 {
+							unrefs = append(unrefs, parentInfo)
+						}
+					} else if removeStartingFromIndex < len(pblock.IPtrs) {
+						// Make sure we're modifying a copy of the
+						// block by fetching it again with blockWrite.
+						pblock, _, err = fd.getter(
```

Comment:
Because `pblock` might already be a modified copy, from another iteration of the loop.  Using `getter` guarantees we're fetching that copy if it already exists and has been placed in the dirty cache.